cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java

```diff
@@ -274,45 +274,51 @@ private void maybeIncreaseTopicReplicationFactor(ZkUtils zkUtils,
                  + " added in none rack-aware way.");
       } else {
-        throw new RuntimeException("Unable to increase topic " + topic + " replica factor to " + replicationFactor
+        throw new RuntimeException("Unable to increase topics " + topics + " replica factor to " + replicationFactor
                                    + " since there are only " + brokersByRack.size() + " racks in the cluster.");
       }
     }
 
     scala.collection.mutable.Map<TopicAndPartition, Seq<Object>> newReplicaAssignment = new scala.collection.mutable.HashMap<>();
-    List<String> racks = new ArrayList<>(brokersByRack.keySet());
-    int [] cursors = new int[racks.size()];
-    int rackCursor = 0;
-    for (MetadataResponse.PartitionMetadata pm : topicMetadata.partitionMetadata()) {
-      if (pm.replicas().size() < replicationFactor) {
-        List<Object> newAssignedReplica = new ArrayList<>();
-        Set<String> currentOccupiedRack = new HashSet<>();
-        // Make sure the current replicas are in new replica list.
-        pm.replicas().forEach(node -> {
-          newAssignedReplica.add(node.id());
-          currentOccupiedRack.add(rackByBroker.get(node.id()));
-        });
-        // Add new replica to partition in rack-aware(if possible), round-robin way.
-        while (newAssignedReplica.size() < replicationFactor) {
-          if (!currentOccupiedRack.contains(racks.get(rackCursor)) || currentOccupiedRack.size() == racks.size()) {
-            String rack = racks.get(rackCursor);
-            int cursor = cursors[rackCursor];
-            newAssignedReplica.add(brokersByRack.get(rack).get(cursor));
-            currentOccupiedRack.add(rack);
-            cursors[rackCursor] = (cursor + 1) % brokersByRack.get(rack).size();
+    for (String topic : topics) {
+      MetadataResponse.TopicMetadata topicMetadata = AdminUtils.fetchTopicMetadataFromZk(JavaConversions.asScalaSet(Collections.singleton(topic)),
+                                                                                         zkUtils,
+                                                                                         ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT)).head();
+      List<String> racks = new ArrayList<>(brokersByRack.keySet());
+      int[] cursors = new int[racks.size()];
+      int rackCursor = 0;
+      for (MetadataResponse.PartitionMetadata pm : topicMetadata.partitionMetadata()) {
+        if (pm.replicas().size() < replicationFactor) {
+          List<Object> newAssignedReplica = new ArrayList<>();
+          Set<String> currentOccupiedRack = new HashSet<>();
+          // Make sure the current replicas are in new replica list.
+          pm.replicas().forEach(node -> {
+            newAssignedReplica.add(node.id());
+            currentOccupiedRack.add(rackByBroker.get(node.id()));
+          });
+          // Add new replica to partition in rack-aware(if possible), round-robin way.
```

Comment:
could we use `AdminUtil. assignReplicasToBrokers()` for rack aware assignment?

Reply:
IIUC, `AdminUtil. assignReplicasToBrokers()` does not consider the current replica distribution, but here we do not want to move current replicas, we just want to add new replica in Rack-aware way to satisfy replication factor requirement. 

we actually use this `AdminUtil. assignReplicasToBrokers()` to increase the partition count(if needed) in this Class,  because there we can assign wherever we want.

Reply:
@alexwang08 has this issue been resolved?